### PR TITLE
Add `Base64DecodingStream`

### PR DESCRIPTION
**Pull request by andrewlock**
## Summary of changes

Adds `Base64DecodingStream` helper class

## Reason for change

In a few places (particularly Remote Config and Newtonsoft.JSON) we have code like this:

```csharp
string someValue = // 
var contentDecode = Convert.FromBase64String(someValue);
using var stream = new MemoryStream(contentDecode);
```

This takes an existing `string`, decodes it from base64 to a `byte[]`, then feeds that `byte[]` into a `MemoryStream`. For big strings, that's a potentially large extra array allocation we can avoid. Instead, `Base64DecodingStream` acts effectively as a `MemoryStream` over the `string`, doing the decode either on the fly (.NET Core), or by using a pooled buffer to decode the input string in chunks.

## Implementation details

Basically asked 🤖 Claude to do this, then reviewed and tidied up and iterated. The general idea is:

- Keep track of where in the string we are
- **.NET Framework/.NET Standard only**
  - Decode the next section of the string into an array-pool rented buffer, using the vendored `Base64.DecodeFromUtf8InPlace`
  - Copy the buffer into the destination `byte[]`
- **.NET Core only**
  - Decode the next section of the string directly into the destination `Span<T>`

It didn't seem worth trying to unify these two paths, given the lack of `Convert.TryFromBase64Chars` in .NET Framework. We _could_ use `Base64.DecodeFromUtf8` to write directly to the destination span instead, but we would still need to do the narrowing, and doing that in the destination span is a little risky, as it would mean we write a bunch of bytes which are then leftover junk. It's _probably_ still fine, but I don't know that it's worth the complexity/risk.

> [!WARNING]
> Rather than handle the case where you're passed a destination buffer that's <3 bytes, this implementation currently just throws. Otherwise we have to decode into a stackallocated buffer, and hang onto the "overflow" bytes to avoid returning 0 when we're not EOF, which is a bit of a pain. For the places we currently use it, I don't think this will be a problem, but if others disagree, we can handle the edge case too.

## Test coverage

Added a variety of unit tests for the implementation

## Other details

https://datadoghq.atlassian.net/browse/LANGPLAT-940

This will be part of a Remote config stack, but for now kept it agnostic

### CODE REVIEW

File: tracer/src/Datadog.Trace/Util/Streams/Base64DecodingStream.cs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+        // All valid base64 characters are in the ASCII range (0–127),
+        // so we can safely narrow each char to a byte for the UTF-8 decoder.
+        var buf = _buffer;
+        var str = _base64;
+        var start = _charPosition;
+
+        for (var i = 0; i < charsToProcess; i++)
+        {
+            buf[i] = (byte)str[start + i];
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Util/Streams/Base64DecodingStream.cs]: On the one hand, definitely an edge case, but on the other, probably best to address it nevertheless

File: tracer/test/Datadog.Trace.Tests/Util/Base64DecodingStreamTests.cs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+            int bytesRead;
+            while ((bytesRead = await stream.ReadAsync(buffer, 0, buffer.Length)) > 0)
+            {
+                ms.Write(buffer, 0, bytesRead);
+            }
+
+            return ms.ToArray();
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/test/Datadog.Trace.Tests/Util/Base64DecodingStreamTests.cs:331]: Oh, it _did_, I guess I removed it 😅
